src/psbt/mod.rs

```diff
@@ -861,6 +800,160 @@ impl PsbtExt for Psbt {
+                    let derived = xpk.derive_public_key(&secp)?;
+                    bip32_derivation.borrow_mut().insert(
+                        derived.to_public_key().inner,
+                        (xpk.master_fingerprint(), xpk.full_derivation_path()),
+                    );
+                    Ok(derived)
+                })
+                .map_err(UpdateInputError::DerivationError)?;
+            inp.bip32_derivation = bip32_derivation.into_inner();
+
+            match &derived {
+                Descriptor::Bare(_) | Descriptor::Pkh(_) | Descriptor::Wpkh(_) => {}
+                Descriptor::Sh(sh) => match sh.as_inner() {
+                    descriptor::ShInner::Wsh(wsh) => {
+                        inp.witness_script = Some(wsh.inner_script());
+                        inp.redeem_script = Some(wsh.inner_script().to_v0_p2wsh());
+                    }
+                    descriptor::ShInner::Wpkh(..) => inp.redeem_script = Some(sh.inner_script()),
+                    descriptor::ShInner::SortedMulti(_) | descriptor::ShInner::Ms(_) => {
+                        inp.redeem_script = Some(sh.inner_script())
+                    }
+                },
+                Descriptor::Wsh(wsh) => inp.witness_script = Some(wsh.inner_script()),
+                Descriptor::Tr(_) => unreachable!("Tr is dealt with separately"),
+            }
+
+            derived
+        };
+
+        let script_pubkey = derived.script_pubkey();
+        if check_witness_utxo {
+            match &inp.witness_utxo {
+                Some(witness_utxo) => {
+                    if witness_utxo.script_pubkey != script_pubkey {
+                        return Err(UpdateInputError::WitnessUtxoCheck);
+                    }
+                }
+                None => {
+                    if derived.desc_type().is_segwit() {
+                        // we allow non-segwit descriptors to be missing witness_utxo
+                        return Err(UpdateInputError::WitnessUtxoCheck);
+                    }
+                }
+            }
+        }
+
+        if let Some(vout) = check_non_witness_utxo {
+            match &inp.non_witness_utxo {
+                Some(non_witness_utxo) => {
+                    let txout = non_witness_utxo
+                        .output
+                        .get(vout as usize)
+                        .ok_or(UpdateInputError::NonWitnessUtxoCheck)?;
+                    if txout.script_pubkey != script_pubkey {
+                        return Err(UpdateInputError::NonWitnessUtxoCheck);
+                    }
+                }
+                None => {
+                    match derived {
+                        Descriptor::Tr(_) => { /*Tr outputs can always omit non_witness_utxo*/ }
+                        _ => return Err(UpdateInputError::NonWitnessUtxoCheck),
+                    }
+                }
+            }
+        }
+        // only set the input once all the checks have passed
+        *self = inp;
```

Comment:
Can you please make the above into a separate function that only operates on `inp`. 

```
    fn update_with_descriptor(
        &mut self,
        ...
        )  -> ...
 {
 	let mut inp = self.clone();
 	update_with_descriptor_helper(&mut inp, ...);
 	*self = inp;
 }
```
This way we can be sure that we don't accidentally mutate `self` in the function body in the future. 
 
We should also avoid the allocation here, but I can do that might be complicated code-wise. I will perhaps address that in future as a low-priority issue.

Reply:
I removed the allocation with a helper function that both the checked and unchecked version use in: 8c8f82b895e94ec0cbb4a5eeb0ab15b94be64ead

The idea is that the helper optionally checks the SPK is correct before it mutates anything.